adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java

```diff
@@ -195,80 +192,356 @@ public void run() {
-                        }
-
-                        key = key.trim();
-                        value = StringExtensions.removeQuoteInHeaderValue(value.trim());
-
-                        if (headerItems.containsKey(key)) {
-                            Logger.w(TAG + methodName,
-                                    "Key/value pair list contains redundant key. ",
-                                    "Redundant key: " + key,
-                                    ADALError.DEVELOPER_BEARER_HEADER_MULTIPLE_ITEMS);
-                        }
-
-                        headerItems.put(key, value);
-                    } else {
-                        // invalid format
-                        throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
-                    }
+        }
+
+        final List<Challenge> challenges = Challenge.parseChallenges(authenticateHeader);
+
+        // Grab the Bearer challenge
+        Challenge bearerChallenge = null;
+
+        for (final Challenge challenge : challenges) {
+            if (challenge.getScheme().equalsIgnoreCase(BEARER)) {
+                bearerChallenge = challenge;
+                break;
+            }
+        }
+
+        if (null != bearerChallenge) {
+            final Map<String, String> challengeParams = bearerChallenge.getParameters();
+            String authority = challengeParams.get(AUTHORITY_KEY);
+            String resource = challengeParams.get(RESOURCE_KEY);
+
+            if (StringExtensions.isNullOrBlank(authority)) {
+                throw new ResourceAuthenticationChallengeException(AUTH_HEADER_MISSING_AUTHORITY);
+            }
+
+            // Remove wrapping quotes (if present)
+            authority = authority.replaceAll("^\"|\"$", "");
+
+            if (StringExtensions.isNullOrBlank(authority)) {
+                throw new ResourceAuthenticationChallengeException(AUTH_HEADER_MISSING_AUTHORITY);
+            }
+
+            if (!StringExtensions.isNullOrBlank(resource)) {
+                resource = resource.replaceAll("^\"|\"$", "");
+            }
+
+            return new AuthenticationParameters(authority, resource);
+        }
+
+        throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
+    }
+
+    /**
+     * An authentication challenge.
+     * Format follows <a href="https://tools.ietf.org/html/rfc7235#section-4.1">RFC-7235</a>.
+     *
+     * @see <a href-"https://tools.ietf.org/html/rfc7617">RFC-7617</a>
+     * @see <a href="https://tools.ietf.org/html/rfc6750">RFC-6750</a>
+     */
+    private static class Challenge {
+
+        /**
+         * Regex sequence intended to be prefixed with another value. Whichever value precedes it
+         * will be parsed/grouped-out, assuming it is not bounded by a par of double-quotes ("").
+         */
+        private static final String REGEX_UNQUOTED_LOOKAHEAD = "(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)";
```

Comment:
Pardon me if I miss some context here. So we are changing the entire regex here? The regex here is different from the dev branch. What will happen if the challenge begin with schema "Bearer"? By calling parseScheme, right?

Reply:
The previous regex required that the `WWW-Authenticate` header begin with `Bearer` and not contain other challenges. The new parser should be tolerant of an arbitrary number of challenges and parse them. The `Bearer` challenge is only sought at the end to return the correct `AuthenticationParameters` object